### PR TITLE
xen_4_8: init at 4.8.1

### PR DESCRIPTION
**Pull request by michalpalka**
This commit adds the xen_4_8 package to be used instead of
xen (currently at 4.5.5):
 * Add packages xen_4_8, xen_4_8-slim and xen_4_8-light
 * Add packages qemu_xen_4_8 and qemu_xen_4_8-light to be used
   with xen_4_8-slim and xen_4_8-light respectively.
 * Add systemd to buildInputs of xen (it is required by oxenstored)
 * Adapt xen service to work with the new version of xen
 * Use xen-init-dom0 to initlilise dom0 in xen-store
 * Currently, the virtualisation.xen.stored option is ignored
   if xen 4.8 is used

###### Motivation for this change

This introduces Xen 4.8.1, which is the most current version, along with a few needed changes and a few fixes.

To use the new xen, use these configuration options.
```
virtualisation.xen.package = pkgs.xen_4_8;
virtualisation.xen.qemu-package = pkgs.xen_4_8;
virtualisation.xen.qemu = "${pkgs.xen_4_8}/lib/xen/bin/qemu-system-i386";
```

`xen_4_8-slim` may be used with this configuration:
```
virtualisation.xen.package = pkgs.xen_4_8-slim;
virtualisation.xen.qemu-package = pkgs.qemu_xen_4_8;
virtualisation.xen.qemu = "${pkgs.qemu_xen_4_8}/bin/qemu-system-i386";
```

Finally, `xen_4_8-light` may be used with this configuration:

```
virtualisation.xen.package = pkgs.xen_4_8-light;
virtualisation.xen.qemu-package = pkgs.qemu_xen_4_8-light;
virtualisation.xen.qemu = "${pkgs.qemu_xen_4_8-light}/bin/qemu-system-i386";
```

One notable limitation of this PR is that the `virtualisation.xen.stored` is ignored with xen 4.8, which I hope to address in the future.

Furthermore, there are a few statements in `xen-dom0.nix` and `generic.nix` that are conditional on the xen version, which is not ideal. Please let me know if you would like it to be handled differently.

Also, the package refers to the fork https://github.com/michalpalka/xen-libhvm of the `xen-libhvm` repo, which needed to be adapted to work with xen 4.8. This might equally be solved using a patch.

Please let me know if you have any feedback.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/virtualization/xen/0004-fix-oxenstored-systemd-notification-race.patch [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+--- xen/tools/hotplug/Linux/launch-xenstore.in	2017-05-27 10:19:32.000000000 +0200
++++ xen-delay1/tools/hotplug/Linux/launch-xenstore.in	2017-05-29 23:33:53.000000000 +0200
+@@ -66,6 +66,11 @@
+ 
+ 	systemd-notify --booted 2>/dev/null || timeout_xenstore $XENSTORED || exit 1
+ 
++	# We wait for 10 seconds here as otherwise the sd_notify() notification
++	# from the oxenstored process might not be properly registered by systemd
++	# if this process quits before the notification is sent.
++	sleep 10
```
Comment (Mic92) [inline pkgs/applications/virtualization/xen/0004-fix-oxenstored-systemd-notification-race.patch]: 10 seconds sounds like a lot to me for a message processed by systemd. Is there something else, which slows down the process?

Comment (michalpalka) [inline pkgs/applications/virtualization/xen/0004-fix-oxenstored-systemd-notification-race.patch]: It is hard to pick the right value here, as the call to `sd_notify()` may be delayed by an arbitrary amount of time. Probably 1 second would be enough, but I chose 10 just to be on a safe side. Please note that this delay will not delay the boot process, as systemd will go ahead and start other units as soon as it gets the notification. So the only effect of a longer delay will be the shell script running for some time afterwards. On the other hand, if the notification is lost then the startup of the unit will hang for a longer time until it gets timed out by systemd and fails.

This is an ugly workaround–I believe that you can solve it properly by using `MAINPID=` together with `sd_notify()`. I will submit a bug report about it to xen.

**Comment by oxij** [comment]
Awesome! Thanks! Will test on my machines in the next couple of days.

Problems I see without testing:

* You introduce tabs in xen-dom0.nix (and I can now assume your tabstop = 8 :) )
* The systemd patch is ugly. It potentially slows down complete system
  boot by 10s. Is it upstream somewhere? Can we get it from upstream?
  If not, how do they solve the same problem upstream?
* Can 4.8's oxenstored build without systemd dependency? (A problem for
  SLNOS which uses s6 init).

**Comment by michalpalka** [comment]
Hi!

> Problems I see without testing:
>
> * You introduce tabs in xen-dom0.nix (and I can now assume your tabstop = 8 :) )

That's an oversight–I'll fix that and amend the commit
> * The systemd patch is ugly. It potentially slows down complete system
>   boot by 10s. Is it upstream somewhere? Can we get it from upstream?
>   If not, how do they solve the same problem upstream?

It does not slow the boot by 10s, because systemd continues as soon as it gets the notification from oxenstored. Interestingly, as I just found out, the patch is not needed anymore as I cannot reproduce the problematic behaviour with systemd 233. It seems that there was a problem with systemd 232, which led to the notification being dropped. Given that, it seems that the scripts in xen are correct, and the patch can be dropped–I'll drop it from the commit.
> * Can 4.8's oxenstored build without systemd dependency? (A problem for
>   SLNOS which uses s6 init).

Yes, oxenstored was previously built without an explicit systemd dependency. As a result, systemd was included when the `xen` package was compiled, but not `xen-slim` or `xen-light`. However, if oxenstored is built without systemd, it cannot be used with systemd services, so systemd needs to be somehow included conditionally.

**Comment by michalpalka** [comment]
I have just amended the commit removing the tabs and the race condition workaround. I have not done anything about the systemd dependency, and I am not sure how to continue. Should the packages have a `useSystemd` flag?

**Comment by oxij** [comment]
> As a result, systemd was included when the `xen` package was compiled, but not `xen-slim` or `xen-light`.

Eh? Why? Systemd wasn't even referenced, IIRC.

> I have just amended the commit removing the tabs and the race condition workaround. I have not done anything about the systemd dependency, and I am not sure how to continue. Should the packages have a `useSystemd` flag?

Nah, don't bother, I'll do that after I test it with s6 later.

LGTM. Thanks again!

**Comment by michalpalka** [comment]
> Eh? Why? Systemd wasn't even referenced, IIRC.
It seems that systemd was propagated from one of the other dependencies, and just picked up by the configure script.

**Comment by oxij** [comment]
@grahamc can you merge this?

**Comment by michalpalka** [comment]
I would like to add security patches against this version of xen. Should I amend this PR, or create a new one?

**Comment by oxij** [comment]
I would add security patches on top here (but as separate commits, no squashing).

**Comment by michalpalka** [comment]
I have added the commit with security patches, and rebased everything.

File: pkgs/applications/virtualization/xen/0004-fix-oxenstored-systemd-notification-race.patch (hunk 1 above)
Comment (0xABAB) [inline pkgs/applications/virtualization/xen/0004-fix-oxenstored-systemd-notification-race.patch]: The sleep 10 is fragile, as pointed out by others. If it cannot be done in any other way, leave it like this, but if it can be fixed, fix it. What happened to the bug report to Xen?

File: nixos/modules/virtualisation/xen-dom0.nix [hunk 2]
```diff
@@ -267,26 +267,36 @@ in
         mkdir -p /var/lib/xen # so we create them here unconditionally.
         grep -q control_d /proc/xen/capabilities
         '';
-      serviceConfig.ExecStart = ''
-        ${cfg.stored}${optionalString cfg.trace " -T /var/log/xen/xenstored-trace.log"} --no-fork
-        '';
+      serviceConfig = if cfg.package.version < "4.8" then
+	{ ExecStart = ''
+	    ${cfg.stored}${optionalString cfg.trace " -T /var/log/xen/xenstored-trace.log"} --no-fork
```
Comment (0xABAB) [inline nixos/modules/virtualisation/xen-dom0.nix]: Please do not hardcode /var/log/xen. As a default it's fine.

File: nixos/modules/virtualisation/xen-dom0.nix [hunk 3]
```diff
@@ -306,14 +316,17 @@ in
...
       preStart = ''
         mkdir -p /var/run/xen
         ${optionalString cfg.trace "mkdir -p /var/log/xen"}
         grep -q control_d /proc/xen/capabilities
         '';
       serviceConfig = {
         ExecStart = ''
-          ${cfg.package}/bin/xenconsoled${optionalString cfg.trace " --log=all --log-dir=/var/log/xen"}
+          ${cfg.package}/bin/xenconsoled\
+            ${optionalString ((cfg.package.version >= "4.8")) " -i"}\
```
Comment (0xABAB) [inline nixos/modules/virtualisation/xen-dom0.nix]: The constant "4.8" is repeated multiple times.

File: pkgs/applications/virtualization/xen/4.8.nix [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    "firmware/ovmf-dir-remote" = {
+      src = fetchgit {
+        url = https://xenbits.xen.org/git-http/ovmf.git;
+        rev = "173bf5c847e3ca8b42c11796ce048d8e2e916ff8";
+        sha256 = "07zmdj90zjrzip74fvd4ss8n8njk6cim85s58mc6snxmqqv7gmcr";
+      };
+      meta.description = "Xen's fork of OVMF";
+    };
+  } // {
+    # TODO: patch Xen to make this optional?
```
Comment (0xABAB) [inline pkgs/applications/virtualization/xen/4.8.nix]: If this is important, create an issue for this.

File: pkgs/applications/virtualization/xen/0004-fix-oxenstored-systemd-notification-race.patch (hunk 1 above)
Comment (michalpalka) [inline pkgs/applications/virtualization/xen/0004-fix-oxenstored-systemd-notification-race.patch]: Actually, this code has since been dropped from the PR–see comment below, which I reproduce here:

> Interestingly, as I just found out, the patch is not needed anymore as I cannot reproduce the problematic behaviour with systemd 233. It seems that there was a problem with systemd 232, which led to the notification being dropped. Given that, it seems that the scripts in xen are correct, and the patch can be dropped–I'll drop it from the commit.

File: nixos/modules/virtualisation/xen-dom0.nix (hunk 2 above)
Comment (michalpalka) [inline nixos/modules/virtualisation/xen-dom0.nix]: The code that you are referring to has been carried over from the service that was created for xen 4.5 (or earlier). An argument for keeping it is to try to minimise the change. If you insist, then I can remove it, but otherwise we may keep it as is and drop it when we drop xen 4.5.

File: nixos/modules/virtualisation/xen-dom0.nix (hunk 3 above)
Comment (michalpalka) [inline nixos/modules/virtualisation/xen-dom0.nix]: Would you see a named constant here instead?

File: pkgs/applications/virtualization/xen/4.8.nix (hunk 4 above)
Comment (michalpalka) [inline pkgs/applications/virtualization/xen/4.8.nix]: This is also carried over from the previous package. I'm not sure if I understand this TODO. Is the idea to use ipxe packaged separately?

File: nixos/modules/virtualisation/xen-dom0.nix (hunk 2 above)
Comment (michalpalka) [inline nixos/modules/virtualisation/xen-dom0.nix]: Looking at the rest of the module, `/var/log/xen` seems to be hardcoded in a lot of other places. If you want to get rid of it, then I suggest to do it with a separate PR.

**Comment by oxij** [comment]
michalpalka <notifications@github.com> writes:

> Is the idea to use ipxe packaged separately?

Yes, to make xen-light even lighter.

> Actually, this code has since been dropped from the PR–see comment below, which I reproduce here:

No, I just fetched this branch and its still there.
Also the tabs in xen-dom0.nix are still there.

**Comment by michalpalka** [comment]
>> Is the idea to use ipxe packaged separately?
>
>Yes, to make xen-light even lighter.
I'll create an issue.
>> Actually, this code has since been dropped from the PR–see comment below, which I reproduce here:
>
>No, I just fetched this branch and its still there.
It seems that I have dropped the reference to the patch, but not the patch itself–sorry for that, I will fix it.
> Also the tabs in xen-dom0.nix are still there.
That's right–I'll fix it too.

**Comment by michalpalka** [comment]
I have pushed new versions of commits, which hopefully address all concerns.
> It seems that I have dropped the reference to the patch, but not the patch itself–sorry for that, I will fix it.

Actually, the reference to the patch was also there. I must have pushed old (unamended) commits by mistake. Thanks for the vigilance.

**Comment by michalpalka** [comment]
Thanks a lot @Mic92 @oxij @0xABAB @joachifm for reviewing and merging!

It would be great if some of you could also take a look at #25724, which adds pvgrub that is useful to have with xen.

File: nixos/modules/virtualisation/xen-dom0.nix (hunk 3 above)
Comment (0xABAB) [inline nixos/modules/virtualisation/xen-dom0.nix]: Yes, just to make sure people don't make search and replace errors.
